### PR TITLE
Cast system

### PR DESCRIPTION
**Pull request by Milice**
Hopefully this cast system will be acceptable.

### CODE REVIEW

**Comment by login12** [comment]
Awesome!!!!
Great job man!
Thanks alot for this contribute

**Comment by login12** [comment]
I got this crash.

`[Error - mysql_real_query] Query: UPDATE `players_online` SET `cast_password` = ''on'' WHERE `player_id` = '949';
Message: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'on'' WHERE `player_id` = '949'' at line 1
pure virtual method called
terminate called without an active exception
Aborted`


:)

File: src/protocolgame.cpp [hunk 1]
```diff
@@ -837,6 +847,134 @@ void ProtocolGame::parseLookInBattleList(NetworkMessage& msg)
...
 
+void ProtocolGame::parseExecuteCommand(const std::string& text) {
+	size_t pos = text.find(' ');
+	std::string command = text.substr(1, pos - 1);
+	if (command.empty()) {
+		return;
+	}
+
+	if(command == "commands") {
+		player->sendChannelMessage("", "Available commands:", TALKTYPE_CHANNEL_O, CHANNEL_CAST, false);
```
Comment (djarek) [inline src/protocolgame.cpp]: Data race on the player object.

File: src/protocolspectator.cpp [hunk 2]
```diff
@@ -0,0 +1,1085 @@
...
+		case 0x67:
+		case 0x68:
+		case 0x6A:
+		case 0x6B:
+		case 0x6C:
+		case 0x6D: 
+			if(recvbyte == 0x64) {
+				sendCancelWalk();
+			}
+			sendMoveCreature(player, player->getPosition(), player->getTile()->getStackposOfCreature(spectator, player->getCreature()), player->getPosition(), player->getTile()->getStackposOfCreature(spectator, player->getCreature()), false);
```
Comment (djarek) [inline src/protocolspectator.cpp]: Data race on the player object.

File: src/player.h [hunk 3]
```diff
@@ -105,7 +105,7 @@ struct Skill {
 
 typedef std::map<uint32_t, uint32_t> MuteCountMap;
 
-static constexpr int32_t PLAYER_MAX_SPEED = 1500;
+static constexpr int32_t PLAYER_MAX_SPEED = 5000;
```
Comment (djarek) [inline src/player.h]: ?

File: src/player.h [hunk 4]
```diff
@@ -1260,6 +1300,7 @@ class Player final : public Creature, public Cylinder
 		int32_t offlineTrainingSkill = -1;
 		int32_t offlineTrainingTime = 0;
 		int32_t idleTime = 0;
+		uint32_t spectatorCount = 0;
```
Comment (djarek) [inline src/player.h]: What's the point of this variable?

File: src/player.h [hunk 5]
```diff
@@ -1182,6 +1218,9 @@ class Player final : public Creature, public Cylinder
 		std::map<uint32_t, int32_t> storageMap;
 
 		std::vector<OutfitEntry> outfits;
+		std::vector<ProtocolSpectator*> spectators;
```
Comment (djarek) [inline src/player.h]: Storing raw pointers to objects managed by shared_ptr is a bad idea IMO.

File: src/protocolgame.cpp [hunk 6]
```diff
@@ -865,6 +1003,20 @@ void ProtocolGame::parseSay(NetworkMessage& msg)
 		return;
 	}
 
+	if(channelId == CHANNEL_CAST && player->isLiveCasting()) {
+		InstantSpell* instantSpell = g_spells->getInstantSpell(text);
```
Comment (djarek) [inline src/protocolgame.cpp]: Data race on g_spells.

File: src/protocolgame.cpp [hunk 7]
```diff
@@ -865,6 +1003,20 @@ void ProtocolGame::parseSay(NetworkMessage& msg)
 		return;
 	}
 
+	if(channelId == CHANNEL_CAST && player->isLiveCasting()) {
```
Comment (djarek) [inline src/protocolgame.cpp]: Data race on the player object.

File: src/protocolspectator.cpp [hunk 8]
```diff
@@ -0,0 +1,1085 @@
...
+void ProtocolSpectator::release()
+{
+	if (spectator && spectator->client == shared_from_this()) {
+		if(player->isLiveCasting()) {
+			std::stringstream ss;
+			ss << spectator->getName() << " has left the cast.";
+			player->sendChannelMessage("", ss.str(), TALKTYPE_CHANNEL_O, CHANNEL_CAST);
+		}
+		spectator->client.reset();
+		spectator = nullptr;
```
Comment (djarek) [inline src/protocolspectator.cpp]: Memory leak of spectator?

File: src/player.h (hunk 4 above)
Comment (Milice) [inline src/player.h]: Variable removed. Thank you for pointing this out.

File: src/player.h (hunk 5 above)
Comment (Milice) [inline src/player.h]: Modified to store a shared_ptr, thanks.

File: src/player.h (hunk 3 above)
Comment (Milice) [inline src/player.h]: Fixing this, forgot to revert from my local test.

File: src/protocolgame.cpp (hunk 1 above)
Comment (Milice) [inline src/protocolgame.cpp]: Thank you for pointing this out, do you have any tip on how i SHOULD handle these things to not data race? Thank you.

**Comment by totolol123** [comment]
What is the differences related to @djarek cast system?

**Comment by DamianDEB** [comment]
Error compiling  'getLevelPercent': is not a member of 'Player'
and problem to add table on phpmyadmin
 CREATE TABLE IF NOT EXISTS `players_online` (
   `player_id` int(11) NOT NULL,
   `cast_password` varchar(30) DEFAULT NULL,
   `cast_spectators` smallint(5) NOT NULL DEFAULT '0'
   PRIMARY KEY (`player_id`)
 ) ENGINE=MEMORY;

**Comment by login12** [comment]
any progress? bugs has been fixeds?

**Comment by DamianDEB** [comment]
login12 
better ways use otx with works cast and some new system like a unjustified frags window 
minimal work fix and have good engine tfs is so far to be good 
https://github.com/mattyx14/otxserver/tree/otxserv3/path_10_9

File: src/player.cpp [hunk 9]
```diff
@@ -73,6 +73,44 @@ Player::~Player()
...
+	if(!password.empty()) {
+		castPassword = password;
+	}
+
+	sendChannel(CHANNEL_CAST, "Live Cast", nullptr, nullptr);
+
+	Database* db = Database::getInstance();
+
+	std::ostringstream query;
+	query << "UPDATE `players_online` SET `cast_password` = '" << db->escapeString(password) << "' WHERE `player_id` = '" << getGUID() <<"';";
```
Comment (DamianDEB) [inline src/player.cpp]: yes in !cast (password) get error in console in !cast and connect to spect get  debug and crash server this code its crappy ...

**Comment by Eternal-Scripts** [comment]
I found a problem, if u enter in cast with 2+ ppl and u have monsters/creatures in screen he debug.

(i think is anything type move creature idk).

**Comment by DamianDEB** [comment]
Eternal-Scripts  on tfs 1,2 or 1,3? on my 1.2 this system crash after spec connect to caster

**Comment by Fluppix** [comment]
Working good with normal client but cause crash if I login as spectator with OTClient. Anyone can confirm this?

**Comment by MatheusGrilo** [comment]
I don't know if you are still working on this, but:
1st open a live cast;
2nd log in this cast as spectator;
3rd close the client of spectator (X button).

TFS Crash.

**Pull request closed without merging**